src/wp-includes/class-wp-block-type.php

```diff
@@ -295,15 +295,19 @@ public function __construct( $block_type, $args = array() ) {
 	 *
 	 * @param string $name Deprecated property name.
 	 *
-	 * @return string|null|void The value read from the new property if the first item in the array provided,
-	 *                          null when value not found, or void when unknown property name provided.
+	 * @return string|string[]|null|void The value read from the new property if the first item in the array provided,
+	 *                                   null when value not found, or void when unknown property name provided.
 	 */
 	public function __get( $name ) {
 		if ( ! in_array( $name, $this->deprecated_properties ) ) {
 			return;
 		}
 
 		$new_name = $name . '_handles';
+
+		if ( count( $this->{$new_name} ) > 1 ) {
+			return $this->{$new_name};
```

Comment:
The challenge here is that returning an array here means, the value won't be compatible with the REST API endpoint:

https://github.com/WordPress/wordpress-develop/blob/04e8bb4bb5ed37dade1b1ddda634e45f7503820d/src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php#L696-L733

One way to fix that would be to return only the first item in the REST API endpoint somewhere after:

https://github.com/WordPress/wordpress-develop/blob/04e8bb4bb5ed37dade1b1ddda634e45f7503820d/src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php#L261-L267

This is why we had simplified handling before.

Reply:
Ah, good catch, thank you!

> One way to fix that would be to return only the first item in the REST API endpoint

Would that match the previous behavior of the endpoint (before #3108)?

It's a bit unfortunate that it'd be inconsistent with `WP_Block_Type` which would return the entire array for `editor_script`, but as long as it matches the previous behavior, it should be fine I guess.

Reply:
The best way to confirm would be to play with the following unit test:

https://github.com/WordPress/wordpress-develop/blob/755d5baff3fde70ebd2faf433a0e0cb46dd12a56/tests/phpunit/tests/rest-api/rest-block-type-controller.php#L206L270

My guess is that it will ignore the array and return `null` for deprecated fields.

Reply:
Thank you for the pointer! Playing with it now. 

> My guess is that it will ignore the array and return `null` for deprecated fields.

Yeah, that's exactly what happens. 

<details>
<summary>
Diff with test case
</summary>

```diff
diff --git a/tests/phpunit/tests/rest-api/rest-block-type-controller.php b/tests/phpunit/tests/rest-api/rest-block-type-controller.php
index ed3ee62fe8..9e5255f71b 100644
--- a/tests/phpunit/tests/rest-api/rest-block-type-controller.php
+++ b/tests/phpunit/tests/rest-api/rest-block-type-controller.php
@@ -338,6 +338,23 @@ class REST_Block_Type_Controller_Test extends WP_Test_REST_Controller_Testcase {
                $this->assertNull( $data['style'] );
        }
 
+       /**
+        * @ticket 56733
+        */
+       public function test_get_item_deprecated() {
+               $block_type = 'fake/deprecated';
+               $settings   = array(
+                       'editor_script' => array( 'guten', 'berg' ),
+               );
+               register_block_type( $block_type, $settings );
+               wp_set_current_user( self::$admin_id );
+               $request  = new WP_REST_Request( 'GET', '/wp/v2/block-types/' . $block_type );
+               $response = rest_get_server()->dispatch( $request );
+               $data     = $response->get_data();
+               $this->assertSameSets( $settings['editor_script'], $data['editor_script_handles'] );
+               $this->assertSameSets( $settings['editor_script'], $data['editor_script'] );
+       }
+
        public function test_get_variation() {
                $block_type = 'fake/variations';
                $settings   = array(
```
</details>

I'll try to repro what the behavior was before https://github.com/WordPress/wordpress-develop/pull/3108.

In the meantime: Could it be an option to make the schema for the deprecated fields more permissive?

```diff
diff --git a/src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php b/src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php
index af29924c76..a8c7a0c476 100644
--- a/src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php
+++ b/src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php
@@ -697,7 +697,7 @@ class WP_REST_Block_Types_Controller extends WP_REST_Controller {
                $deprecated_properties      = array(
                        'editor_script' => array(
                                'description' => __( 'Editor script handle. DEPRECATED: Use `editor_script_handles` instead.' ),
-                               'type'        => array( 'string', 'null' ),
+                               'type'        => array( 'string', 'array', 'null' ),
                                'default'     => null,
                                'context'     => array( 'embed', 'view', 'edit' ),
                                'readonly'    => true,
```

(And so on for the other deprecated fields.) 
This would make the test pass.

Reply:
In addition to that, it's also why we soft-deprecated old fields in favor of their replacement.

Reply:
(I've pushed 588323a for some unit test coverage.)

Reply:
Okay, I went with returning the first array item after all. The reason is that we're now storing everything in an array (e.g. `editor_script_handles`), and if there's only one element, we cannot differentiate if it was assigned from an array like that, or if it came from a plain string. This means that we would return that string here; but `null` if there are more than one items in the array. But that would be pretty inconsistent behavior.